{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8741390",
   "metadata": {},
   "source": [
    "# Finding Mutated Motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23886369",
   "metadata": {},
   "source": [
    "Given a string s and a motif t, an alignment of a substring of s against all of t is called a fitting alignment. Our aim is to find a substring sâ€² of s that maximizes an alignment score with respect to t\n",
    "\n",
    ".\n",
    "\n",
    "Note that more than one such substring of s\n",
    "\n",
    "may exist, depending on the particular strings and alignment score used. One candidate for scoring function is the one derived from edit distance; In this problem, we will consider a slightly different alignment score, in which all matched symbols count as +1 and all mismatched symbols (including insertions and deletions) receive a cost of -1. Let's call this scoring function the mismatch score. See Figure 1 for a comparison of global, local, and fitting alignments with respect to mismatch score.\n",
    "\n",
    "Given: Two DNA strings s\n",
    "and t, where s has length at most 10 kbp and t\n",
    "\n",
    "represents a motif of length at most 1 kbp.\n",
    "\n",
    "Return: An optimal fitting alignment score with respect to the mismatch score defined above, followed by an optimal fitting alignment of a substring of s\n",
    "against t. \n",
    "If multiple such alignments exist, then you may output any one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccb78590",
   "metadata": {},
   "outputs": [],
   "source": [
    "x=''\n",
    "header=''\n",
    "y=''\n",
    "fasta = {}\n",
    "with open('rosalind_sims_869_7_dataset.txt','r') as f :\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue\n",
    "        if line.startswith(\">\"):\n",
    "            header = line[1:]\n",
    "            if header not in fasta:\n",
    "                fasta[header] = []\n",
    "            continue\n",
    "        sequence = line\n",
    "        fasta[header].append(sequence)\n",
    "v=list(fasta)\n",
    "for i in fasta[v[0]]:\n",
    "    x=x+i\n",
    "for i in fasta[v[1]]:\n",
    "    y=y+i\n",
    "\n",
    "\n",
    "\n",
    "x = list(x)\n",
    "\n",
    "y=list(y)\n",
    "\n",
    "m = len(x)\n",
    "n = len(y)\n",
    "map = []\n",
    "for i in range(m+1):\n",
    "    map.append([])\n",
    "    for j in range(n+1):\n",
    "        map[i].append(0)\n",
    "for i in range (1,n+1):\n",
    "    map[0][i]=-i*1\n",
    "\n",
    "for i in range (1,m+1):\n",
    "    for j in range (1,n+1):\n",
    "\n",
    "        if x[i-1]==y[j-1]:\n",
    "            map[i][j]=map[i-1][j-1]+1\n",
    "        else:\n",
    "            map[i][j]=max(map[i-1][j-1]-1,map[i-1][j]-1,map[i][j-1]-1)\n",
    "g=0\n",
    "\n",
    "for i in range (m+1):\n",
    "    for j in range (n+1):\n",
    "        if map[i][j] > g and j==n :\n",
    "            g=map[i][j]\n",
    "            q = i\n",
    "            p = j\n",
    "            break\n",
    "\n",
    "len(y)\n",
    "i=q\n",
    "j=p\n",
    "\n",
    "xans=''\n",
    "yans=''\n",
    "while   j != 0 :\n",
    "    if x[i-1]==y[j-1]:\n",
    "        xans = x[i - 1] + xans\n",
    "        yans = y[j - 1] + yans\n",
    "        i=i-1\n",
    "        j=j-1\n",
    "        continue\n",
    "    elif map[i-1][j-1]-1==map[i][j]:\n",
    "        xans = x[i - 1] + xans\n",
    "        yans = y[j - 1] + yans\n",
    "        i = i - 1\n",
    "        j = j - 1\n",
    "        continue\n",
    "    elif map[i-1][j]-1==map[i][j]:\n",
    "        xans = x[i - 1] + xans\n",
    "        yans = '-' + yans\n",
    "        i=i-1\n",
    "        continue\n",
    "    elif map[i][j-1]-1==map[i][j]:\n",
    "        xans = '-' + xans\n",
    "        yans = y[j - 1] + yans\n",
    "        j=j-1\n",
    "\n",
    "\n",
    "print(g)\n",
    "print(xans)\n",
    "print(yans)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
